### PR TITLE
Refine performance evaluation method

### PR DESCRIPTION
**Pull request by cyy0714**
1. Precision and recall calculation is now based on the top metric. The original calculation take the average of downstram metrics.

2. Function replay takes the time and duration instead of a fix start and end time.

3. ReplayRunnable will clone anomalies before the monitoring window.

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobResource.java [hunk 1]
```diff
@@ -579,20 +582,34 @@ public Response evaluateAlertFilterByAutoTuneId(@PathParam("autotuneId") long id
   @POST
   @Path("replay/function/{id}")
   public Response anomalyFunctionReplay(@PathParam("id") @NotNull long functionId,
-      @QueryParam("start") @NotNull String replayStartTimeIso, @QueryParam("end") @NotNull String replayEndTimeIso,
+      @QueryParam("time") String replayTimeIso, @QueryParam("duration") @DefaultValue("30") int replayDuration,
```
Comment (bl44) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobResource.java]: Does this mean that you will fix the end to now? This is less flexible than the original one. Why making this choice?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/performanceEvaluation/PrecisionByTimePerformanceEvaluation.java [hunk 2]
```diff
@@ -52,13 +52,15 @@ public double evaluate() {
...
-      }
-      avgPrecision += (double) dimensionToOverlapTimeLength.get(dimensions) / dimensionToDetectedAnomalyTimeLength.get(dimensions);
+      totalDetectedAnomalyTimeLength += dimensionToDetectedAnomalyTimeLength.get(dimensions);
+      totalDimensionOverlapTimeLength += dimensionToOverlapTimeLength.get(dimensions);
+    }
+    if (totalDetectedAnomalyTimeLength == 0) {
+      return Double.NaN;
     }
-    return avgPrecision / ((double) dimensionToDetectedAnomalyTimeLength.size());
+    return (double) totalDimensionOverlapTimeLength / totalDetectedAnomalyTimeLength;
```
Comment (bl44) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/performanceEvaluation/PrecisionByTimePerformanceEvaluation.java]: Is your definition of F1 score need to change too?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/lib/FunctionReplayRunnable.java [hunk 3]
```diff
@@ -64,7 +74,6 @@ public FunctionReplayRunnable(DetectionJobScheduler detectionJobScheduler, Anoma
     setForceBackfill(isForceBackfill);
     setTuningParameter(tuningParameter);
     setFunctionAutotuneConfigId(functionAutotuneConfigId);
-    setSpeedUp(true);
```
Comment (bl44) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/lib/FunctionReplayRunnable.java]: So there will be no SpeedUp?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobResource.java (hunk 1 above)
Comment (cyy0714) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobResource.java]: No. User can assign different start time in its query. My modification here is to set-up a default start time.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/performanceEvaluation/PrecisionByTimePerformanceEvaluation.java (hunk 2 above)
Comment (cyy0714) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomalydetection/performanceEvaluation/PrecisionByTimePerformanceEvaluation.java]: It is not required. The definition of F1 score is not changed. It is a different calculation strategy among different sub-metrics.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/lib/FunctionReplayRunnable.java (hunk 3 above)
Comment (cyy0714) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/lib/FunctionReplayRunnable.java]: The default speedup value is set in the base constructor. This value here is a duplicate.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobResource.java [hunk 4]
```diff
@@ -579,20 +582,34 @@ public Response evaluateAlertFilterByAutoTuneId(@PathParam("autotuneId") long id
...
+      @QueryParam("durationUnit") String durationUnit,
```
Comment (cecilynie) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobResource.java]: If we have a default duration value, why we don't have a default duration unit?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/lib/FunctionReplayRunnable.java [hunk 5]
```diff
@@ -118,22 +127,23 @@ public void run() {
       functionName.append(entry.getValue());
     }
     try {
-      clonedFunctionId = onboardResource.cloneAnomalyFunctionById(tuningFunctionId, functionName.toString(), false);
-      this.lastClonedFunctionId = clonedFunctionId;
+      clonedFunctionId = onboardResource.cloneAnomalyFunctionById(tuningFunctionId, functionName.toString(), cloneAnomaly);
+      this.clonedFunctionId = clonedFunctionId;
```
Comment (cecilynie) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/lib/FunctionReplayRunnable.java]: The clonedFunctionId will always be overrided to the new clonedFunctionId, and in the end you only keep the very last cloned function Id. Shall we keep the name of "lastClonedFunctionId" or some what change to a better name to specify this logic?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/lib/FunctionReplayRunnable.java [hunk 6]
```diff
@@ -176,14 +188,16 @@ public void run() {
 
     // clean up and kill itself
     if(selfKill) {
-      onboardResource.deleteExistingAnomalies(Long.toString(clonedFunctionId), replayStart.getMillis(),
+      onboardResource.deleteExistingAnomalies(Long.toString(clonedFunctionId), 0,
           replayEnd.getMillis());
       anomalyFunctionDAO.deleteById(clonedFunctionId);
```
Comment (cecilynie) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/lib/FunctionReplayRunnable.java]: If is selfKipp, clonedFunctionId is also being deleted. What about the clonedFunctionId contained in Runnable?

**Pull request merged**